lib/gc/Transforms/LowerPackAndUnpack.cpp

```diff
@@ -0,0 +1,74 @@
+//===-- LowerPackAndUnpack.cpp - Lower PackOp and UnpackOp ------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "gc/Transforms/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+namespace mlir {
+namespace gc {
+#define GEN_PASS_DEF_LOWERPACKANDUNPACK
+#include "gc/Transforms/Passes.h.inc"
+} // namespace gc
+} // namespace mlir
+
+namespace {
+class LowerPackAndUnpack
+    : public mlir::gc::impl::LowerPackAndUnpackBase<LowerPackAndUnpack> {
+
+  void runOnOperation() override {
+    auto *ctx = &getContext();
+    IRRewriter rewriter(ctx);
+
+    getOperation()->walk([&](tensor::PackOp target) {
+      OpBuilder::InsertionGuard g(rewriter);
+      rewriter.setInsertionPoint(target);
+
+      FailureOr<linalg::LowerPackResult> res =
+          linalg::lowerPack(rewriter, target);
+      if (failed(res)) {
+        llvm::dbgs()
+            << "cannot lower tensor.pack to pad + expand + transpose\n";
+        return signalPassFailure();
+      }
+    });
+
+    getOperation()->walk([&](tensor::UnPackOp target) {
+      OpBuilder::InsertionGuard g(rewriter);
+      rewriter.setInsertionPoint(target);
+
+      FailureOr<linalg::LowerUnPackOpResult> res =
+          linalg::lowerUnPack(rewriter, target);
```

Comment:
Seems this function is already there, just missing a good calling point in our pipeline. I think it's not necessary to make it as a standalone pass, it's better to do this when lowering linalg to arith/math.

Reply:
I guess one of the advantages of making this as a standalone pass is: we can further optimize multiple packs/unpacks before lowering them to the bufferizable ops. There might be cases like 2D->pack->4D -> pack -> VNNI format, can we just merge these two `pack` and leaving only one chain of `pad+reshape+transpose` instead of two?

Reply:
> do this when lowering linalg to arith/math

There is no such one single pass for now, so I put the proposed pass in `populateTensorPasses`. The proposed pass actually replaces `tensor.pack/unpack` ops with bufferizable operations in `tensor` and `linalg` dialects, which makes it not a strictly **lowering** pass.

Reply:
> merge these two pack and leaving only one chain of pad+reshape+transpose instead of two

It's may be better to keep the pass single-purposed and leave the fusion related features to the fusion pass.

Reply:
> It's may be better to keep the pass single-purposed and leave the fusion related features to the fusion pass.

It's not like the general fusion pass, but instead of an elimination step for redundant pack ops.

Reply:
> I guess one of the advantages of making this as a standalone pass is: we can further optimize multiple packs/unpacks before lowering them to the bufferizable ops. There might be cases like 2D->pack->4D -> pack -> VNNI format, can we just merge these two `pack` and leaving only one chain of `pad+reshape+transpose` instead of two?

It sounds like a DAG lowering instead of single-op lowering. Since there's no good DAG pattern match in mlir right now, we can do this as a separate loop and refactor this when DAG pattern match support is added.

Reply:
> merge these two pack

As far as I can see, only one `pack` op can not express the semantics of `2D->pack->4D -> pack -> VNNI format`, although one chain of `pad+reshape+transpose` can. So just as Zhennan said, it's a DAG lowering instead of single-op lowering.

Reply:
Since the purpose of this pass is to lower the `tensor.pack/unpack` ops to bufferizable ops, is it better to move it to `populateBufferizationPasses`?